CRM/PivotData/DataLeave.php

```diff
@@ -0,0 +1,566 @@
+      $fieldParts = explode('.', $field);
+      if ($fieldParts[0] === $entityName) {
+        $result[] = $fieldParts[1];
+      }
+    }
+
+    return $result;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  protected function formatResult($data, $dataKey = null, $level = 0) {
+    $result = [];
+    $resultRow = [];
+
+    foreach ($data as $key => $dataRow) {
+      $leaveDataRow = $this->getRowValues($dataRow, 'LeaveRequest');
+      $leaveDatesData = $dataRow['api.LeaveRequestDate.get']['values'];
+      $absenceTypeRow = $this->getRowValues(array_shift($dataRow['api.AbsenceType.get']['values']), 'AbsenceType');
+      $contractData = $dataRow['api.HRJobContract.get']['values'];
+      $contactDataRow = $this->getRowValues(array_shift($dataRow['api.Contact.get']['values']), 'Contact');
+      foreach($leaveDatesData as $leaveDate) {
+        $leaveDateRow = $this->getRowValues($leaveDate, 'LeaveRequestDate');
+        $leaveBalanceRow = $this->getRowValues(array_shift($leaveDate['api.LeaveBalanceChange.get']['values']), 'LeaveBalanceChange');
+        $hasContract = false;
+        foreach($contractData as $contract) {
+          if ($leaveDate['date'] >= $contract['period_start_date'] && (empty($contract['period_end_date']) || $leaveDate['date'] <= $contract['period_end_date'])) {
+            $hasContract = true;
+            $currentRevision = $this->getAPIEntityData('HRJobContractRevision', ['jobcontract_id' => $contract['id']], 'getcurrentrevision');
+            $contractRow = $this->getRowValues($contract, 'HRJobContract');
+            $jobRolesRow = $this->getRowValues(array_shift($contract['api.HrJobRoles.get']['values']), 'HrJobRoles');
+
+            $jobPayData = $this->getAPIEntityData('HRJobPay', ['jobcontract_revision_id' => $currentRevision['pay_revision_id']]);
+            $jobHourData = $this->getAPIEntityData('HRJobHour', ['jobcontract_revision_id' => $currentRevision['hour_revision_id']]);
+            $jobPensionData = $this->getAPIEntityData('HRJobPension', ['jobcontract_revision_id' => $currentRevision['pension_revision_id']]);
+            $jobPayRow = $this->getRowValues(array_shift($jobPayData), 'HRJobPay');
+            $jobHourRow = $this->getRowValues(array_shift($jobHourData), 'HRJobHour');
+            $jobPensionRow = $this->getRowValues(array_shift($jobPensionData), 'HRJobPension');
+
+            $row = array_merge($leaveDataRow, $absenceTypeRow, $contactDataRow,
+              $leaveDateRow, $leaveBalanceRow, $contractRow, $jobRolesRow, $jobPayRow, $jobHourRow, $jobPensionRow);
+            $resultRow[] = $this->formatRow($key, $row);
+          }
+        }
+
+        if(!$hasContract) {
+          $row = array_merge($leaveDataRow, $absenceTypeRow, $contactDataRow, $leaveDateRow, $leaveBalanceRow);
+          $resultRow[] = $this->formatRow($key, $row);
+        }
+      }
+
+      $headers = array_merge($this->emptyRow, $this->additionalHeaderFields);
+      $finalRow = [];
```

Comment:
Well, I discovered that some API's will not return a field when the value of that field is null, e.g the JobContract API will not return the period_end_date field when it is null for a contract. In the final row array, I try to assign a null value the field that was not returned by checking the data in resultRow, If this is not done, this will cause some misalignment in the header and column values in the pivot table as some value that is meant for a header will appear for another column header.

Reply:
I think this is not something specific to some APIs, it's just the normal behavior of the CiviCRM APIs. In this case, I think it's a better idea to extract this logic to a separate method with a name and docblock that better describe what is happening here and why you need this.

Reply:
Good Idea.